doc/design.md

```diff
@@ -0,0 +1,1268 @@
+
+This document describes a proposed new library, to become part of the Mantid project.
+The working title of the project that led to the design proposed in the following has been **Workspace-2.0**, i.e., the set goal was to find a way to improve upon the workspace types that are currently available in Mantid.
+For reasons that will become clear below we adopt the title `Dataset` instead of the old working title.
+
+Before starting the design and prototyping phase, [comments about the current implementation and requirements for the new design](https://github.com/mantidproject/documents/blob/master/Project-Management/CoreTeam/workspace-notes.md) had been gathered.
+For a big picture of the current situation, consider the class inheritance diagram for the existing workspaces in Mantid:
+
+![figure: current inheritance diagram for workspaces in Mantid](workspace-hierarchy.png)
+
+To put this diagram in perspective, note that one major point of criticism of the current implementation was "not enough workspace types", i.e., for the redesign an even larger variety was requested.
+Based on these requirements a design and prototyping process was started in early 2018.
+Initially a couple of different options have been prototyped but it soon became apparent that only one of them was a real option.
+We refer to the [early prototype documentation](https://github.com/mantidproject/workspace-sandbox) for details on the abandoned options and why they were abandoned.
+In agreement with reviewers of those early investigations, effort has thus been mostly focused on the `Dataset` option, described below.
+
+Altogether, as of 2018-10-10, a total of 65 working days (approx 3.5 FTE months) have been invested in the design and prototyping process.
+We also refer to the [prototype development notes](development-notes.md) for more details and motivation of the features presented in the following.
+
+
+## <a name="overview"></a>High level design overview
+
+We propose a new class called `Dataset`, with the capability to replace almost all existing workspaces types while at the same time providing more features and more flexibility.
+The inspiration for `Dataset` comes from the `Dataset` type in [xarray](http://xarray.pydata.org/en/stable/).
+
+Quoting from the website:
+*"**xarray** (formerly **xray**) is an open source project and Python package that aims to bring the labeled data power of pandas to the physical sciences, by providing N-dimensional variants of the core pandas data structures."*
+
+`xarray.Dataset` is roughly speaking a `dict` of `numpy.ndarray` objects with labeled coordinates and dimensions.
+Our design adds features that are not present in `xarray.Dataset` but represent the core of what Mantid provides to neutron science: Handling histograms of (neutron) data, handling of physical units, and propagation of uncertainties.
+Note that the proposed design is *not* a 1:1 copy of `xarray.Dataset`, there are some significant differences.
```

Comment:
Yes that is a good question, I will add this to the document after this review round, for now let me summarize my current view:

- No unit handling in `xarray`.
- No propagation of uncertainties.
- No support of histograms (i.e., we cannot have a bin-edge axis, which is by 1 longer than the other variables in that dimension).
- Maybe not so suitable for event data.
- Performance may be adequate in many situations, but having a C++ library enables us to get the maximum performance is needed.

Probably many of these things could be added to `xarray`, but I feel that the involved complexities imply that the resulting effort is not less than a from-scratch implementation. Furthermore, `xarray` seems to be less than 5 years old with very few active contributors, so building everything on top of it is an additional (albeit probably small) risk.

Reply:
Regarding the performance, it would be nice to do some benchmarking with respect to `xarray` to help justify our own C++ implementation.

Reply:
For very basic benchmarks there will not be much difference: `operator+` is bound by memory bandwidth and we cannot get much faster than what we get from `numpy`, which `xarray` is using internally.

Apart from that, I agree it would be nice to have benchmarks for more complex operations, but where to start and where to stop? I can test non-standard indexing, which as far as I know `xarray` does in Python, but how relevant is that?

If you have a concrete suggestion I am happy to try though!

What else is there to consider? In C++ we can play some tricks to get better performance, e.g., processing our data in small chunks that fit into cache (to overcome the memory bandwidth limit), use custom allocators, memory pools, or object pools to overcome allocation/initialization overheads. I have played with some of these and do see improvements, but it is not included in the design right now (see https://github.com/mantidproject/dataset/tree/vector_pool, https://github.com/mantidproject/dataset/blob/vector_pool/benchmark/dataset_benchmark.cpp#L193, and https://github.com/mantidproject/dataset/blob/vector_pool/src/memory_pool.h if you are interested). I cannot tell whether we will end up using any of these or not, but with a Python-based solution we simply do not have this option.